experimental/dds/tree2/src/feature-libraries/treeIndex.ts

```diff
@@ -0,0 +1,33 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { ChangeAtomId, FieldKey } from "../core";
+import { brand } from "../util";
+
+/**
+ * The tree index records detached field ids and associates them with a change atom ID.
+ */
+export class TreeIndex {
+	private readonly detachedFields = new Map<ChangeAtomId, FieldKey>();
+
+	public constructor(private readonly name: string) {}
+
+	/**
+	 * Returns a field key for the given change atom ID and ID. Should be unique for
+	 * this index.
+	 */
+	public getFieldKey(changeAtomId: ChangeAtomId, id: string): FieldKey {
+		assert(!this.detachedFields.has(changeAtomId), "Detached field already exists");
```

Comment:
I think this assert may need to be removed for the sake of the restore (old name: "revive") codepath.